### PR TITLE
Add metaclass support, support PEP 3115 and PEP 487

### LINKED ISSUE #855: Metaclasses under Python 3
When pondering #850, it occurred to me that I don't know how to set a metaclass in Hy. (It doesn't come up much.) I mean in Python2 we can do it with the old syntax (`__metaclass__ = Meta` in the class body), which is straightforward in Hy. But this syntax isn't allowed in Python3. After reading the docs, I don't think Hy can do this. That or the docs need some updating. It's an issue either way.

### PR DESCRIPTION
**Pull request by vodik**
Fixes #855 (though I didn't think to search for the issue until after I implemented it...)

### CODE REVIEW

File: hy/compiler.py [hunk 1]
```diff
@@ -2071,14 +2071,35 @@ def rewire_init(expr):
...
+            if not isinstance(keywords[0].value, ast.Name):
+                raise HyTypeError(base_list, "specified metaclass must be "
+                                             "a symbol")
+
+            e = HyExpression([
+                    HySymbol("setv"),
+                    HySymbol("__metaclass__"),
+                    HySymbol(keywords[0].value.id)
+                ]).replace(base_list)
+            spoof_positions(e)
```
Comment (Kodiologist) [inline hy/compiler.py]: Do you need both `.replace` and `spoof_positions` here? Shouldn't one be enough?

File: tests/compilers/test_ast.py [hunk 2]
```diff
@@ -217,6 +218,12 @@ def test_ast_bad_defclass():
     cant_compile("(defclass a None None)")
 
 
+@pytest.mark.skipif(PY3, reason="Python 3 supports class keywords")
+def test_ast_bad_defclass_with_kwargs():
+    "Make sure AST can't compile invalid yield-from"
```
Comment (Kodiologist) [inline tests/compilers/test_ast.py]: The comment doesn't match the test.

File: tests/native_tests/test_metaclass.hy [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+(defn test-metaclass []
+  (defclass QuestBase [type]
+     [--new-- (fn [cls name bases classdict]
+               (if (in "__swallow__" classdict)
+                   (assoc classdict "swallow" (get classdict "__swallow__")))
+               (-> (super QuestBase cls) (.--new-- cls name bases classdict)))])
+
+  (defclass Quest [:metaclass QuestBase]
+    [--swallow-- "african"])
```
Comment (Kodiologist) [inline tests/native_tests/test_metaclass.hy]: Why the use of double underscores? `__swallow__` isn't a special Python name, which is what double underscores on both sides of a name are typically reserved for, and it doesn't trigger the automatic Python mangling that a name like `__swallow` would, which we probably don't need for this test, anyway.

Comment (Kodiologist) [reply tests/native_tests/test_metaclass.hy]: If all you want to do is distinguish `swallow` from a similar but distinct variable, how about `swallow_`?

File: hy/compiler.py (hunk 1 above)
Comment (vodik) [reply hy/compiler.py]: Yeah, I thought it was weird when I looked at it, `replace` should be enough and idea. But I copy and pasted the approach and elsewhere both are used:

https://github.com/hylang/hy/blob/10cd36d13d71010c38e8b411ebf13390f80b9542/hy/compiler.py#L414-L424

File: tests/native_tests/test_metaclass.hy (hunk 3 above)
Comment (vodik) [reply tests/native_tests/test_metaclass.hy]: No good reason. Honestly, I'm not married to this particular test case at all. The other two came from their respective PEP's, this one I derived.

File: hy/compiler.py (hunk 1 above)
Comment (Kodiologist) [reply hy/compiler.py]: Such cases might also be redundant, couldn't they? (Not that I'm expecting you to fix them in this PR.)

Comment (vodik) [reply hy/compiler.py]: Yeah, agreed. Just pointing it out so we know. I'm certainly going to investigate it.

File: hy/compiler.py [hunk 4]
```diff
@@ -2071,14 +2071,35 @@ def rewire_init(expr):
...
             base_list = expressions.pop(0)
             if not isinstance(base_list, HyList):
                 raise HyTypeError(base_list, "Base classes must be a list.")
-            bases_expr, bases, _ = self._compile_collect(base_list)
+            bases_expr, bases, keywords = self._compile_collect(base_list, with_kwargs=True)
 
         body = Result()
 
+        if not PY3 and keywords:
+            # Dirty shim for Python 2 to make :metaclass keyword work.
```
Comment (gilch) [inline hy/compiler.py]: I'm not sure if this is worth the added complexity to the compiler. We can already use metaclasses in Python2 using `(setv __metaclass__ foo)`. Our Python2 support is ending in 2020 at the latest. It's fine to fix bugs, but we shouldn't be adding new features for it.

Comment (vodik) [reply hy/compiler.py]: I'd be fine with this if everyone else is.

Comment (Kodiologist) [reply hy/compiler.py]: I can confirm that I would be fine with that.

**Comment by Kodiologist** [comment]
Looks good. Why not put the new tests in `py3_only_tests.hy` and `py36_only_tests.hy` instead of their own files?

**Comment by vodik** [comment]
Just taste. pytest doesn't require that we have to put them all together and I like having the file name reflect what's inside of it.

Can do though, just a sec.

**Comment by Kodiologist** [comment]
It's only two new tests, so it seems a little silly to make two new files. When (or if) the version-specific files get large, then we should start breaking them up, by all means. We already have some test files that are probably too big, like language.hy and core.hy.

**Comment by vodik** [comment]
No worries, I get it.

**Comment by Kodiologist** [comment]
@gilch, do you approve?

File: NEWS.rst [hunk 5]
```diff
@@ -1,5 +1,13 @@
 .. default-role:: code
 
+Unreleased
+==============================
+
+New Features
+------------------------------
+* `defclass` can now take keyword arguments for metaclass for
+  Python 3 metaclass support.
```
Comment (gilch) [inline NEWS.rst]: "for metaclass for Python 3 metaclass support"? I think this could be worded better.

**Comment by vodik** [comment]
Fixed up NEWS... finally

**Comment by Kodiologist** [comment]
@gilch, do you approve?

**Comment by gilch** [comment]
Approved. We still have a NEWS conflict though.

**Pull request merged**